### PR TITLE
FIX Image in summaryfields breaks search

### PR DESCRIPTION
**Pull request by sabina-talipova**
## Description
Fields that pass validation on `hasMethod()` check, but do not have a corresponding field or relations in the DB are excluded from `searchable_fields`.

## Parent issue
- https://github.com/silverstripe/silverstripe-framework/issues/10761

### CODE REVIEW

File: src/ORM/DataObject.php [hunk 1]
```diff
@@ -3724,6 +3724,40 @@ public function onAfterBuild()
...
+            $hasRelation = function () use ($component, $relation) {
+                if ($component instanceof DataObject) {
+                    $manyMany = $component->manyMany();
+                    $hasMany = $component->hasMany();
+                    $hasOne = $component->hasOne();
+                    $belongsTo = $component->belongsTo();
+                    if (array_key_exists($relation, $manyMany)
+                        || array_key_exists($relation, $hasMany)
+                        || array_key_exists($relation, $hasOne)
+                        || array_key_exists($relation, $belongsTo)) {
```
Comment (GuySartorelli) [inline src/ORM/DataObject.php]:
```suggestion
                if ($component instanceof DataObject) {
                    if ($component->getRelationType($relation) !== null) {
```

File: src/ORM/DataObject.php [hunk 2]
```diff
@@ -3724,6 +3724,40 @@ public function onAfterBuild()
...
 
+    private function hasRelations($fieldPath)
+    {
+        $component = $this;
+        foreach (explode('.', $fieldPath ?? '') as $relation) {
+            if (!$component) {
+                return null;
+            }
+            $hasRelation = function () use ($component, $relation) {
+                if ($component instanceof DataObject) {
```
Comment (GuySartorelli) [inline src/ORM/DataObject.php]: If this is false, we don't return any value... should this whole function just be this instead?
```php
return ($component instanceof DataObject) && $component->getRelationType($relation) !== null;
```

File: src/ORM/DataObject.php [hunk 3]
```diff
@@ -3755,7 +3789,7 @@ public function searchableFields()
 
                     if ($schema->fieldSpec($this, $name)) {
                         $fields[] = $name;
-                    } elseif ($this->relObject($spec)) {
+                    } elseif ($this->relObject($spec) && $this->hasRelations($spec)) {
```
Comment (GuySartorelli) [inline src/ORM/DataObject.php]: Under what circumstances does `relObject()` return a value record, but `hasRelations()` returns false?
The PHPDoc for `relObject` says:
> If a relation is blank, this will return null instead.
> If a relation name is invalid (e.g. non-relation on a parent) this
> can throw a LogicException.

In other words, `relObject()` should already return null if the relation doesn't exist.

File: tests/php/ORM/Search/SearchContextTest.php [hunk 4]
```diff
@@ -59,12 +59,18 @@ public function testSearchableFieldsDefaultsToSummaryFields()
         $summaryFields = $obj->summaryFields();
         $expected = [];
         foreach ($summaryFields as $field => $label) {
+            if ($field === 'Image.CMSThumbnail'
+                || $field === 'Image.BackLinks.Count') {
+                continue;
+            }
```
Comment (GuySartorelli) [inline tests/php/ORM/Search/SearchContextTest.php]: Why are we skipping these for this test?

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Haven't tested locally yet but there are some things in the implementation and tests that don't quite make sense to me

File: src/ORM/DataObject.php (hunk 3 above)
Comment (sabina-talipova) [reply src/ORM/DataObject.php]: I have excluded fields like `CMSThumbnail` and `StripThumbnail` from `$searchable_fields`. Since, for example `ImageManipulation` adds these methods to the `File` class and they returns `DBHTMLText` they are not part of any relations and don't have any fields in DB. In the `relObject`  method, they easily passed the hasMethod check and are added to the search fields accordingly, but since they are not fields in the data base, they fail the request. As result `ERROR [Emergency]: Uncaught InvalidArgumentException: "CMSThumbnail" is not a field on class "SilverStripe\Assets\Image"`
I think we mention about this issue [here](https://docs.silverstripe.org/en/4/developer_guides/model/scaffolding/#searchable-fields)

> $searchable_fields will default to use the [$summary_fields config](https://docs.silverstripe.org/en/4/developer_guides/model/scaffolding/#summary-fields) if not defined. This works fine unless your $summary_fields config specifies fields that are not stored in the database.

I don't want to add this checking in `relObject` since we use this method in another places, not for `searchable_fields`, probably for some logic it's acceptable to get `$relation` (that is method, but it is not stored in DB) that passed `hasMethod` check.

File: tests/php/ORM/Search/SearchContextTest.php (hunk 4 above)
Comment (sabina-talipova) [reply tests/php/ORM/Search/SearchContextTest.php]: `CMSThumbnail ` and `Count` are helper methods and don't have fields in the database. See my comment above.

File: src/ORM/DataObject.php [hunk 5]
```diff
@@ -3724,6 +3724,32 @@ public function onAfterBuild()
...
+
+            if (ClassInfo::hasMethod($component, $relation)) {
+                if ($component->hasField($relation)
+                    || (($component instanceof DataObject) && $component->getRelationType($relation) !== null)) {
+                    $component = $component->$relation();
+                } else {
+                    return false;
+                }
+            } else {
+                return true;
```
Comment (sabina-talipova) [inline src/ORM/DataObject.php]: The only purpose of this method is to additionally check that whatever `hasMethod` passes is really a relationship getter and not just a helper method. Therefore, assuming that all other checks are passed in `relObject` method and if they are passed, then `hasRelations ` returns `true` for them. If not passed, then this method will not be called.

File: src/ORM/DataObject.php [hunk 6]
```diff
@@ -3724,6 +3724,32 @@ public function onAfterBuild()
...
+            if (!$component) {
+                return false;
+            }
+
+            if (ClassInfo::hasMethod($component, $relation)) {
+                if ($component->hasField($relation)
+                    || (($component instanceof DataObject) && $component->getRelationType($relation) !== null)) {
+                    $component = $component->$relation();
+                } else {
+                    return false;
```
Comment (sabina-talipova) [inline src/ORM/DataObject.php]: Any `$relation` that does not pass this check is not suitable for further passing to the `search` method.

File: src/ORM/DataObject.php (hunk 1 above)
Comment (sabina-talipova) [reply src/ORM/DataObject.php]: Updated

File: src/ORM/DataObject.php (hunk 2 above)
Comment (sabina-talipova) [reply src/ORM/DataObject.php]: Updated

**Comment by GuySartorelli** [comment]
I think this approach isn't complete.

I have used the following class when testing locally. Note that the `Company` class mentioned is just some random DataObject with a `Name` field.

```php
use SilverStripe\Assets\Image;
use SilverStripe\ORM\DataObject;
use SilverStripe\ORM\FieldType\DBField;

class SearchableFieldsTest extends DataObject
{
    private static $db = [
        'Name' => 'Varchar',
        'Biography' => 'HTMLText',
    ];

    private static $has_one = [
        'Company' => Company::class,
        'ProfileImage' => Image::class
    ];

    private static $many_many  = [
        'PastCompanies' => Company::class
    ];

    private static $summary_fields = [
        'ID',
        'Name',
        'Biography.Plain',
        'CompanyID',
        'Company.Category',
        'Company.Name.Plain',
        'PastCompanies.Name',
        'PastCompanies.Count', // shouldn't be searchable
        'ProfileImage.CMSThumbnail', //  shouldn't be searchable
        'MethodOne', //  shouldn't be searchable
        'MethodTwo', //  shouldn't be searchable
        'Company', //  shouldn't be searchable
    ];

    public function getMethodOne()
    {
        return implode(' ', $this->Company()->Employees()->column('Name'));
    }

    public function MethodTwo()
    {
        return DBField::create_field('Varchar', 'this is the value');
    }
}
```

## Problems with the current implementation

Correctly included:
- Direct fields on the class (e.g. `Name`)
- Direct fields on the class, even if we're calling `Plain` or `Nice` etc on that field (e.g. `Biography.Plain`)
- Direct fields on a has_one relation (e.g. `Company.Category`)
- Direct fields on a many_many or has_many relation (e.g. `PastCompanies.Name`)

Still included, but shouldn't be:
- Methods on the class (e.g. `MethodTwo`)

Not included, but _should_ be:
- Direct fields on a relation if we're calling `Plain` or `Nice` etc on that field (e.g. `Company.Name.Plain`)

Not tested:
- Direct fields on chained relations (e.g. `Company.Customers.Location`)
- Direct fields on chained relations if we're calling `Plain` or `Nice` etc on that field (e.g. `Company.Customers.Location.Plain`)

## Recommendation

Instead of `hasRelations()`, which is still allowing methods and _isn't_ correctly allowing relation fields in some cases, have a method, similar to the below, which takes a dot-syntax field name (e.g. `Name` or `Company.Customers.Location.Plain`) and returns _only_ the part that is backed by the database (e.g. `Name` or `Company.Customers.Location`).

Note that this implementation is based heavily on `relObject()`, but has been changed so that any time we come across something that _cannot_ be backed by the database (e.g. a method call that isn't for getting a relation list), we return null - and otherwise we return the _dot syntax field name_ for the database-backed field, which excludes method calls such as `Plain`.
This works correctly for chained relations.

```php
    private function getDatabaseBackedField(string $fieldPath): ?string
    {
        $component = $this;
        $fieldParts = [];

        foreach (explode('.', $fieldPath ?? '') as $nextPart) {
            if (!$component) {
                return null;
            }
            $fieldParts[] = $nextPart;

            if ($component instanceof Relation || $component instanceof DataList) {
                if ($component->hasMethod($nextPart)) {
                    // If the next part is a method, we don't have a database-backed field.
                    return null;
                }
                // The next part could either be a field, or another relation
                $singleton = DataObject::singleton($component->dataClass());
                if ($singleton->dbObject($nextPart)) {
                    // If the next part is a DBField, we've found the database-backed field.
                    break;
                }
                $component = $component->relation($nextPart);
            } elseif ($component instanceof DataObject && ($component->dbObject($nextPart) instanceof DBField)) {
                // If the next part is a DBField, we've found the database-backed field.
                break;
            } elseif ($component instanceof DataObject && $component->getRelationType($nextPart) !== null) {
                $component = $component->$nextPart();
            } elseif (ClassInfo::hasMethod($component, $nextPart)) {
                // If the next part is a method, we don't have a database-backed field.
                return null;
            } else {
                return null;
            }
        }
        return implode('.', $fieldParts) ?: null;
    }
```

Then, use this instead of everything in the `foreach ($summaryFields as $key => $name) {` block:

```diff
- $spec = $name;
- // Extract field name in case this is a method called on a field (e.g. "Date.Nice")
- if (($fieldPos = strpos($name ?? '', '.')) !== false) {
-     $name = substr($name ?? '', 0, $fieldPos);
- }
- 
- if ($schema->fieldSpec($this, $name)) {
-     $fields[] = $name;
- } elseif ($this->relObject($spec) && $this->hasRelations($spec)) {
-     $fields[] = $spec;
+ if ($field = $this->getDatabaseBackedField($name)) {
+     $fields[] = $field;
 }
```

File: tests/php/Forms/GridField/GridFieldFilterHeaderTest.php [hunk 7]
```diff
@@ -208,8 +208,8 @@ public function testCanFilterAnyColumns()
 
         // test that you cannot filter by non-db field when it falls back to summary_fields
         Config::modify()->remove(Team::class, 'searchable_fields');
-        Config::modify()->set(Team::class, 'summary_fields', ['MySummaryField']);
-        $this->assertFalse($filterHeader->canFilterAnyColumns($gridField));
+        Config::modify()->set(Team::class, 'summary_fields', ['Cheerleader.Name']);
+        $this->assertTrue($filterHeader->canFilterAnyColumns($gridField));
```
Comment (GuySartorelli) [inline tests/php/Forms/GridField/GridFieldFilterHeaderTest.php]: Why have you made this change? The comment above this makes it clear that the intention is to test that you _cannot_ filter by non-db fields... but here you're testing that you _can_ filter by _db fields_

File: tests/php/ORM/DataObjectTest.php [hunk 8]
```diff
@@ -2671,4 +2672,36 @@ public function testDBObjectEnum()
...
+        $this->assertSame('HasOneRelationship.FavouriteTeam', $databaseBackedField);
+
+        $databaseBackedField = $method->invokeArgs($class, ['SubTeams.Comments.Name.Plain']);
+        $this->assertSame('SubTeams.Comments.Name', $databaseBackedField);
+
+        $databaseBackedField = $method->invokeArgs($class, ['ReturnsNull']);
+        $this->assertNull($databaseBackedField);
+
+        $databaseBackedField = $method->invokeArgs($class, ['DynamicField']);
+        $this->assertNull($databaseBackedField);
```
Comment (GuySartorelli) [inline tests/php/ORM/DataObjectTest.php]: Please convert this to use a [dataprovider](https://docs.phpunit.de/en/9.6/writing-tests-for-phpunit.html#writing-tests-for-phpunit-data-providers) - that has a couple of benefits:
- It's a lot easier to see what the various test scenarios are
- It's easier to add new test scenarios
- _all_ of the scenarios will be tested, even if one fails early on, which means we have a clear view of what works and what doesn't

File: tests/php/ORM/DataObjectTest.php [hunk 9]
```diff
@@ -2671,4 +2672,36 @@ public function testDBObjectEnum()
...
+        $this->assertSame('Captain.ShirtNumber', $databaseBackedField);
+
+        $databaseBackedField = $method->invokeArgs($class, ['Captain.FavouriteTeam.ReturnsNull']);
+        $this->assertSame('Captain.FavouriteTeam', $databaseBackedField);
+
+        $databaseBackedField = $method->invokeArgs($class, ['Founder.FavouriteTeam.Captain.ReturnsNull']);
+        $this->assertSame('Founder.FavouriteTeam.Captain', $databaseBackedField);
+
+        $databaseBackedField = $method->invokeArgs($class, ['HasOneRelationship.FavouriteTeam.MyTitle']);
+        $this->assertSame('HasOneRelationship.FavouriteTeam', $databaseBackedField);
```
Comment (GuySartorelli) [inline tests/php/ORM/DataObjectTest.php]: This one should return null - `FavouriteTeam` isn't a database backed field, it's a whole DataObject record.
Any conversation about this one should be on https://github.com/silverstripe/silverstripe-framework/pull/10885#discussion_r1281296765 which provides more context

File: tests/php/ORM/DataObjectTest.php [hunk 10]
```diff
@@ -2671,4 +2672,36 @@ public function testDBObjectEnum()
...
+        $this->assertSame('Captain.FavouriteTeam.Title', $databaseBackedField);
+
+        $databaseBackedField = $method->invokeArgs($class, ['Captain.ShirtNumber']);
+        $this->assertSame('Captain.ShirtNumber', $databaseBackedField);
+
+        $databaseBackedField = $method->invokeArgs($class, ['Captain.FavouriteTeam.ReturnsNull']);
+        $this->assertSame('Captain.FavouriteTeam', $databaseBackedField);
+
+        $databaseBackedField = $method->invokeArgs($class, ['Founder.FavouriteTeam.Captain.ReturnsNull']);
+        $this->assertSame('Founder.FavouriteTeam.Captain', $databaseBackedField);
```
Comment (GuySartorelli) [inline tests/php/ORM/DataObjectTest.php]: This one should return null - `Captain` isn't a database backed field, it's a whole DataObject record.
Any conversation about this one should be on https://github.com/silverstripe/silverstripe-framework/pull/10885#discussion_r1281296765 which provides more context

File: tests/php/ORM/DataObjectTest.php [hunk 11]
```diff
@@ -2671,4 +2672,36 @@ public function testDBObjectEnum()
...
+        $class = new Team([]);
+
+        $databaseBackedField = $method->invokeArgs($class, ['Captain.FavouriteTeam.Title']);
+        $this->assertSame('Captain.FavouriteTeam.Title', $databaseBackedField);
+
+        $databaseBackedField = $method->invokeArgs($class, ['Captain.ShirtNumber']);
+        $this->assertSame('Captain.ShirtNumber', $databaseBackedField);
+
+        $databaseBackedField = $method->invokeArgs($class, ['Captain.FavouriteTeam.ReturnsNull']);
+        $this->assertSame('Captain.FavouriteTeam', $databaseBackedField);
```
Comment (GuySartorelli) [inline tests/php/ORM/DataObjectTest.php]: This one should return null - `FavouriteTeam` isn't a database backed field, it's a whole DataObject record.

The following is meant to handle this scenario:
```php
if ($component->hasMethod($nextPart)) {
    // If the next part is a method, we don't have a database-backed field.
    return null;
}
```

File: tests/php/ORM/DataObjectTest.php [hunk 12]
```diff
@@ -2671,4 +2672,36 @@ public function testDBObjectEnum()
         $vals = ['25.25', '50.00', '75.00', '100.50'];
         $this->assertSame(array_combine($vals ?? [], $vals ?? []), $obj->dbObject('MyEnumWithDots')->enumValues());
     }
+
+    public function testGetDatabaseBackedField()
```
Comment (GuySartorelli) [inline tests/php/ORM/DataObjectTest.php]: Please also include tests for fields directly on the `$class` - such as 'Title' and 'Title.Plain' which should both return 'Title'

Please also include a couple of tests for scenarios where there's _nothing_ there - e.g. "ThisDoesntExist", "Captain.FavouriteTeam.ThisDoesntExist", and "SubTeams.ThisDoesntExist" (testing on the record itself, on a has_one, and on a many relation) which should all return `null`.

File: tests/php/ORM/DataObjectTest.php [hunk 13]
```diff
@@ -2671,4 +2672,36 @@ public function testDBObjectEnum()
...
+    {
+        $dataObjectClass = new DataObject();
+        $method = new ReflectionMethod($dataObjectClass, 'getDatabaseBackedField');
+        $method->setAccessible(true);
+        $class = new Team([]);
+
+        $databaseBackedField = $method->invokeArgs($class, ['Captain.FavouriteTeam.Title']);
+        $this->assertSame('Captain.FavouriteTeam.Title', $databaseBackedField);
```
Comment (GuySartorelli) [inline tests/php/ORM/DataObjectTest.php]: Please also include a test for `'Captain.FavouriteTeam.Title.Plain'`

File: tests/php/ORM/Search/SearchContextTest.php [hunk 14]
```diff
@@ -58,13 +58,24 @@ public function testSearchableFieldsDefaultsToSummaryFields()
         $obj = new SearchContextTest\NoSearchableFields();
         $summaryFields = $obj->summaryFields();
         $expected = [];
-        foreach ($summaryFields as $field => $label) {
+
+        $expectedSearchableFields = [
+            'Name',
+            'Customer.FirstName',
+            'Image',
+            'Image.BackLinks',
```
Comment (GuySartorelli) [inline tests/php/ORM/Search/SearchContextTest.php]: Neither `Image` nor `Image.BackLinks` can be searched against, since they're not _fields_, they're a whole dataobject and a whole HasManyList resepectively.
```suggestion
```

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: tests/php/ORM/DataObjectTest.php (hunk 8 above)
Comment (sabina-talipova) [reply tests/php/ORM/DataObjectTest.php]: Updated

File: tests/php/ORM/DataObjectTest.php (hunk 9 above)
Comment (sabina-talipova) [reply tests/php/ORM/DataObjectTest.php]: Updated

File: tests/php/ORM/DataObjectTest.php (hunk 10 above)
Comment (sabina-talipova) [reply tests/php/ORM/DataObjectTest.php]: Updated

File: tests/php/ORM/DataObjectTest.php (hunk 11 above)
Comment (sabina-talipova) [reply tests/php/ORM/DataObjectTest.php]: Updated

File: tests/php/ORM/DataObjectTest.php (hunk 12 above)
Comment (sabina-talipova) [reply tests/php/ORM/DataObjectTest.php]: Updated

File: tests/php/ORM/DataObjectTest.php (hunk 13 above)
Comment (sabina-talipova) [reply tests/php/ORM/DataObjectTest.php]: Updated

File: tests/php/ORM/Search/SearchContextTest.php (hunk 14 above)
Comment (sabina-talipova) [reply tests/php/ORM/Search/SearchContextTest.php]: Updated

File: tests/php/Forms/GridField/GridFieldFilterHeaderTest.php [hunk 15]
```diff
@@ -205,11 +205,6 @@ public function testCanFilterAnyColumns()
         // this is because we're making a blind assumption it will be filterable later in a SearchContext
         Config::modify()->set(Team::class, 'searchable_fields', ['WhatIsThis']);
         $this->assertTrue($filterHeader->canFilterAnyColumns($gridField));
-
-        // test that you cannot filter by non-db field when it falls back to summary_fields
-        Config::modify()->remove(Team::class, 'searchable_fields');
-        Config::modify()->set(Team::class, 'summary_fields', ['MySummaryField']);
-        $this->assertFalse($filterHeader->canFilterAnyColumns($gridField));
```
Comment (sabina-talipova) [inline tests/php/Forms/GridField/GridFieldFilterHeaderTest.php]: I removed this test case since it passes test of comparison `searchable_fields` & `summary_fields` in `GridFieldFilterHeader::canFilterAnyColumns($gridField)`, since in most cases they will be different. Only if we set identical `searchable_fields` & `summary_fields` with proper DB fields it will go to other step and probably returns `true` since if believe `ArrayList::canFilterBy($by)` , `searchable_field` as key will exist in `GridField->list`.

Comment (GuySartorelli) [reply tests/php/Forms/GridField/GridFieldFilterHeaderTest.php]: If `summary_fields` _only_ has `MySummaryField` (which here is presumably a method name), then falling back to summary_fields should _still_ result in `$filterHeader->canFilterAnyColumns` returning false, because there are _no searchable fields_ (i.e. `MySummaryField` will be excluded, so the set of searchable fields will be empty, so there's nothing to filter by, so `canFilterAnyColumns()` should return false).

Please add this test back in, so that we can be sure that the empty set of searchable fields results in `canFilterAnyColumns` returning false here.

File: src/ORM/DataObject.php [hunk 16]
```diff
@@ -3724,6 +3724,51 @@ public function onAfterBuild()
         $this->extend('onAfterBuild');
     }
 
+    private function getDatabaseBackedField(string $fieldPath): ?string
+    {
+        $component = $this;
+        $fieldParts = [];
+
+        foreach ($parts = explode('.', $fieldPath ?? '') as $nextPart) {
```
Comment (GuySartorelli) [inline src/ORM/DataObject.php]:
```suggestion
        $parts = explode('.', $fieldPath ?? '');
        foreach ($parts as $nextPart) {
```
Please avoid assigning variables inside a foreach clause.

File: src/ORM/DataObject.php [hunk 17]
```diff
@@ -3724,6 +3724,51 @@ public function onAfterBuild()
...
+                    return null;
+                }
+                // The next part could either be a field, or another relation
+                $singleton = DataObject::singleton($component->dataClass());
+                if ($singleton->dbObject($nextPart)) {
+                    // If the next part is a DBField, we've found the database-backed field.
+                    break;
+                }
+                $component = $component->relation($nextPart);
+                array_shift($parts);
```
Comment (GuySartorelli) [inline src/ORM/DataObject.php:3751]: What is the `array_shift` call doing there? Why do we need that? The tests pass without it, so either a test scenario is missing, or this isn't needed.
(the one below _is_ needed for the tests to pass, though)

File: src/ORM/DataObject.php [hunk 18]
```diff
@@ -3724,6 +3724,51 @@ public function onAfterBuild()
...
+            $fieldParts[] = $nextPart;
+            
+            if ($component instanceof Relation || $component instanceof DataList) {
+                if ($component->hasMethod($nextPart)) {
+                    // If the next part is a method, we don't have a database-backed field.
+                    return null;
+                }
+                // The next part could either be a field, or another relation
+                $singleton = DataObject::singleton($component->dataClass());
+                if ($singleton->dbObject($nextPart)) {
```
Comment (GuySartorelli) [inline src/ORM/DataObject.php]:
```suggestion
                if ($singleton->dbObject($nextPart) instanceof DBField) {
```
We should probably be strict here, the same as we are for the `dbObject()` call in the condition below.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
This is shaping up nicely. Just a few small things to clean up now.

File: src/ORM/DataObject.php (hunk 18 above)
Comment (sabina-talipova) [reply src/ORM/DataObject.php]: Updated

File: src/ORM/DataObject.php (hunk 17 above)
Comment (sabina-talipova) [reply src/ORM/DataObject.php:3751]: Updated. New test cases were added.

File: src/ORM/DataObject.php (hunk 16 above)
Comment (sabina-talipova) [reply src/ORM/DataObject.php]: Updated

File: tests/php/Forms/GridField/GridFieldFilterHeaderTest.php (hunk 15 above)
Comment (sabina-talipova) [reply tests/php/Forms/GridField/GridFieldFilterHeaderTest.php]: Updated.

File: tests/php/ORM/DataObjectTest.php [hunk 19]
```diff
@@ -2671,4 +2672,48 @@ public function testDBObjectEnum()
...
+            ['HasOneRelationship.FavouriteTeam.MyTitle', null],
+            ['SubTeams.Comments.Name.Plain', 'SubTeams.Comments.Name'],
+            ['Title', 'Title'],
+            ['Title.Plain', 'Title'],
+            ['DatabaseField', 'DatabaseField'],
+            ['DatabaseField.MethodDoesNotExist', 'DatabaseField'],
+            ['ReturnsNull', null],
+            ['DynamicField', null],
+            ['SubTeams.ParentTeam.Fans', null],
+            ['SubTeams.ParentTeam.Founder.FoundingTeams', null],
```
Comment (sabina-talipova) [inline tests/php/ORM/DataObjectTest.php:2702]: These two cases to cover `array_shift` in case `if ($component instanceof Relation || $component instanceof DataList)`.

**Review by GuySartorelli — APPROVED** [review]
Looks good, works well locally, and provides a much nicer default behaviour than the previous behaviour.
